### PR TITLE
Animated textures.

### LINKED ISSUE #137: Pseudorandom texturing
Use a pseudo-randomly selected frame from animated textures to render e.g. torches.

### PR DESCRIPTION
**Pull request by alexhliu**
Added a time slider which increments the animated texture. Animated textures are now based on block position.

Closes #137, related to #889

### CODE REVIEW

**Comment by leMaik** [comment]
[As already said](https://github.com/chunky-dev/chunky/issues/889#issuecomment-820783636) I'd suggest making `time` a time in milliseconds and changing it to `long`. This will allow adding the more complicated animated texture features (e.g. interpolation, custom frame times, …) later or even animate water.

File: chunky/src/java/se/llbit/chunky/ui/render/AdvancedTab.java [hunk 1]
```diff
@@ -137,6 +140,14 @@ public PictureExportFormat fromString(String string) {
     cacheResolution.onValueChange(value -> {
       scene.sky().setSkyCacheResolution(value);
     });
+    animationTime.setName("Current animation time");
+    animationTime.setTooltip("Current animation time in seconds.");
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/ui/render/AdvancedTab.java:144]:
```suggestion
    animationTime.setTooltip("Current animation time in seconds, used for animated textures.");
```

File: chunky/src/java/se/llbit/chunky/model/FireModel.java [hunk 2]
```diff
@@ -38,10 +40,14 @@
...
               new Vector4(1, 0, 0, 1)),};
 
-  public static boolean intersect(Ray ray, AnimatedTexture[] texture) {
+  public static boolean intersect(Ray ray, AnimatedTexture[] texture, double time) {
     boolean hit = false;
     ray.t = Double.POSITIVE_INFINITY;
-    int i = 0xF & (ray.getCurrentData() >> BlockData.LILY_PAD_ROTATION);
+    Vector3 position = new Vector3(ray.o);
+    position.scaleAdd(Ray.OFFSET, ray.d);
+    int i = (0xF & (ray.getCurrentData() >> BlockData.LILY_PAD_ROTATION))
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/model/FireModel.java:48]: `i` used to be `[0, 15]`

File: chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java [hunk 3]
```diff
@@ -319,6 +308,11 @@
    */
   public int previewCount;
 
+  /**
+   * Current time in seconds. Adjusts animated blocks like fire.
+   */
+  private double animationTime = 0;
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java:314]: Saving and loading this value from the scene description file should also be added

File: chunky/src/java/se/llbit/chunky/model/FireModel.java (hunk 2 above)
Comment (alexhliu) [reply chunky/src/java/se/llbit/chunky/model/FireModel.java:48]: `i` dosen't have to be in the range 0,15. The animation frame used = `i % numFrames`.

**Review by leMaik — APPROVED** [review]

**Comment by leMaik** [comment]
Looking forward to see the other animated texture features and maybe render a small video one day 🤩 

We need to somehow port this to the model refactoring branch. I'll look into that this weekend.

**Pull request merged**